src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java

```
@@ -0,0 +1,610 @@
+            final var feedback = new Feedback();
+            feedback.setPositive(testCase.getErrors() == null && testCase.getFailures() == null);
+            feedback.setText(testCase.getName());
+            String errorMessage = null;
+            // If we have errors or failures, they will always be of length == 1 since JUnit (and the format itself)
+            // should generally only report he first failure in a test case
+            if (testCase.getErrors() != null) {
+                errorMessage = testCase.getErrors().get(0).getMessage();
+            }
+            else if (testCase.getFailures() != null) {
+                errorMessage = testCase.getFailures().get(0).getMessage();
+            }
+            // The assertion message can be longer than the allowed char limit, so we shorten it here if needed.
+            if (errorMessage != null && errorMessage.length() > FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS) {
+                errorMessage = errorMessage.substring(0, FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS);
+            }
+            feedback.setDetailText(errorMessage);
+
+            return feedback;
+        }).collect(Collectors.toList());
+
+        result.setHasFeedback(true);
+        result.addFeedbacks(feedbacks);
+    }
+
+    private TestResults fetchLatestBuildResultFromJenkins(ProgrammingExerciseParticipation participation) {
+        final var projectKey = participation.getProgrammingExercise().getProjectKey();
+        final var planKey = participation.getBuildPlanId();
+        final var url = Endpoint.TEST_RESULTS.buildEndpoint(JENKINS_SERVER_URL.toString(), projectKey, planKey).build(true);
+
+        return restTemplate.getForObject(url.toUri(), TestResults.class);
+    }
+
+    @Override
+    public List<Feedback> getLatestBuildResultDetails(Result result) {
+        // TODO since this is unused as of now
+        return null;
```

Should this throw a "not implemented" exception instead of returning null?

---

Yeah, you are right. I changed it